Дрожжин Иван  gb.ru/lessons/344116/homework

IvanDrojjin opened the pull request:
Дрожжин Иван
Здравствуйте
...........
Для проведения исследований с генеалогическим древом.
Идея: описать некоторое количество компонент, например:
модель человека и дерева

..........
Спасибо.

IvanDrojjin committed:
Create README.md

Добавлена "Инструкция на то как сделать PR"

IvanDrojjin committed:
add gb.ru/lessons/344116/homework 

https://gb.ru/lessons/344116/homework	до 26 июля, 20:00 +03:00 UTC
	Объектно-ориентированное программирование (семинары)
Урок 1. Принципы ООП: Инкапсуляция, наследование, полиморфизм
Реализовать, с учетом ооп подхода, приложение.
Для проведения исследований с генеалогическим древом.
Идея: описать некоторое количество компонент, например:
модель человека и дерева
Под “проведением исследования” можно понимать например получение всех детей выбранного человека.
Более детальное описание проекта и как его реализовать обсуждали в конце семинара
Сделать PR к проекту: https://github.com/Liberate520/homeWork
В качестве ответа указать ссылку на PR
Ссылка на то как сделать пулреквест смотри в материалах к уроку
Если PR все таки не дается, то можно и ссылкой на гит репозиторий

Liberate521 commented on Human.java:
```diff
@@ -0,0 +1,161 @@
+import java.io.*; // Serializable
+
+public class Human implements Serializable {
+    private String fName;
```
f это типа field? Нет необходимости, только засоряете имена переменных

IvanDrojjin committed:
Add files interface

FileReadWrite.java
ObjIO.java

IvanDrojjin committed:
Change files to use interface...

Change  to use interface...
main.java
Human.java
HumanTree.java

IvanDrojjin committed:
Add sort by Name or Age

Реализовать интерфейс Iterable для дерева.
Создать методы сортировки списка людей перед выводом, например по имени или по дате рождения (не менее 2)

Liberate520 commented on HumanTree.java:
```diff
@@ -0,0 +1,117 @@
+import java.io.*; // Serializable
+
+
+public class HumanTree<E extends Human> implements Serializable, Iterable<E> {  // E inheritance Human
```
Зависимость лучше делать от интерфейса

Liberate520 commented on HumanTree.java:
```diff
@@ -0,0 +1,117 @@
+    }
+    
+
+    public String saveToFile(String aFileName)   throws IOException, ClassNotFoundException   {
+         //Сериализация в файл с помощью класса ObjectOutputStream
+        ObjIO aIO = new ObjIO();
+        aIO.writeFile(aFileName, (Serializable) fHumensTree);
+         
+        // ObjectOutputStream objectOutputStream = new ObjectOutputStream(
+        //         new FileOutputStream(aFileName));
+        
+        // objectOutputStream.writeObject(fHumensTree);
+        // objectOutputStream.close();
+        
+        fFileName = aFileName;
+        return "HumanTree save to " + aFileName;
+    }
+    
+    public List <E> loadFromFile(String aFileName)  throws IOException, ClassNotFoundException    {
+        List <E> aHT;
+        // Востановление из файла с помощью класса ObjectInputStream
+        ObjIO aIO = new ObjIO();
+        aHT = (List <E>) aIO.readFile(aFileName);
+        fHumensTree = aHT;
+        fFileName = aFileName;
+        return aHT;
```
Методов по работе с файловой системой в классе дерева быть не должно

IvanDrojjin committed:
add animal.java

IvanDrojjin committed:
add animal.java

IvanDrojjin committed:
add animal.java

IvanDrojjin committed:
add Animal.java

IvanDrojjin committed:
add animal.java

IvanDrojjin committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	Animal.java

IvanDrojjin committed:
add packages

IvanDrojjin committed:
add packages & MVP

IvanDrojjin commented:
@IvanDrojjin
[add packages & MVP](https://github.com/Liberate520/homeWork/pull/694/commits/3abf1fec6f1e3378e5c95b545bdcaeda6961ac40)

Liberate520 commented on src/presenter/Presenter.java:
```diff
@@ -0,0 +1,84 @@
+
+public class Presenter {
+    private View view;
+    private HumanTree<Human> humansTreeConnect;
+    private ObjIO serialize;
```
Создать сервис для объединения работы модели

Liberate520 commented on src/presenter/Presenter.java:
```diff
@@ -0,0 +1,84 @@
+    private HumanTree<Human> humansTreeConnect;
+    private ObjIO serialize;
+
+    private HumanComparatorByAge sortAge;
```
Не характеризует презентер

Liberate520 commented on src/presenter/Presenter.java:
```diff
@@ -0,0 +1,84 @@
+
+    public void sortByName() {
+        humansTreeConnect.sortByName();
+        view.printAnswer(humansTreeConnect.toString());
```
Нарушение принципа DRY. у вас есть метод getHumansInfo

Liberate520 commented on AgeCalculator.java:
```diff
@@ -0,0 +1,15 @@
+// даты и изменять их, добавляя и отнимая необходимое количество дней/месяцев/лет.
+// https://metanit.com/java/tutorial/12.3.php
+
+public class AgeCalculator {
```
Удалить лишние классы из проекта

Liberate520 commented on src/presenter/Presenter.java:
```diff
@@ -0,0 +1,84 @@
+
+    public void addHuman(String name, Gender aSex, LocalDate aBD) {  // String name, String sex, int age
+        // humansTreeConnect.addHuman(new Human(name, sex, age));
+        humansTreeConnect.addHuman(new Human(name, aSex, aBD));
```
презентер не должен заниматься вопросами создания новых людей. Это нарушение первого принципа солид. Перенести работу в класс сервис

Liberate520 commented on src/model/HumanTree.java:
```diff
@@ -0,0 +1,116 @@
+    }
+    
+
+    public String saveToFile(String aFileName) {  // throws IOException, ClassNotFoundException
+         //Сериализация в файл с помощью класса ObjectOutputStream
+        ObjIO aIO = new ObjIO();
+        aIO.writeFile(aFileName, (Serializable) humensTree);
+
+        fileName = aFileName;
+        return "HumanTree save to " + aFileName;
+    }
+    
+    public List <E> loadFromFile(String aFileName) { //  throws IOException, ClassNotFoundException
+        List <E> aHT;
+        // Востановление из файла с помощью класса ObjectInputStream
+        ObjIO aIO = new ObjIO();
+        aHT = (List <E>) aIO.readFile(aFileName);
+        humensTree = aHT;
+        fileName = aFileName;
+        return aHT;
+    }
+    
```
Нарушение первого принципа солид. Работа с файловой системой не является задачей данного класса

Liberate520 commented on src/model/HumanTree.java:
```diff
@@ -0,0 +1,116 @@
+        return aHT;
+    }
+    
+    public List <E> addFromFile(String aFileName)  throws IOException, ClassNotFoundException    {
+        List <E> aHT;
+        // Востановление из файла с помощью класса ObjectInputStream
+        ObjIO aIO = new ObjIO();
+        aHT = (List <E>) aIO.readFile(aFileName);
+        for (E h: aHT){
+            this.addHuman(h);
+        }  
+        fileName = aFileName;
+        return humensTree;
+    }
```
тоже самое
